internal/instance/resource_instance.go

```diff
@@ -942,22 +925,140 @@ func ToProfileListType(ctx context.Context, profiles []string) (types.List, diag
 	return types.ListValueFrom(ctx, types.StringType, profiles)
 }
 
-// networkStateRefreshFunc returns function that refreshes instance's status.
-func instanceStateRefreshFunc(server lxd.InstanceServer, instanceName string) retry.StateRefreshFunc {
-	return func() (any, string, error) {
+// startInstance starts an instance with the given name. It waits for its
+// status to become ready and for number of processes to be more then 0.
+// This way we ensure the instance is both running and successfully
+// connected to the LXD server.
+func startInstance(ctx context.Context, server lxd.InstanceServer, instanceName string, refInterval time.Duration) diag.Diagnostic {
+	st, etag, err := server.GetInstanceState(instanceName)
+	if err != nil {
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to retrieve state of instance %q", instanceName), err.Error())
+	}
+
+	// Return if already running.
+	if st.Status == api.Running.String() && st.Processes > 0 {
+		return nil
+	}
+
+	startReq := api.InstanceStatePut{
+		Action:  "start",
+		Force:   false,
+		Timeout: utils.ContextTimeout(ctx, 3*time.Minute),
+	}
+
+	// Start the instance.
+	op, err := server.UpdateInstanceState(instanceName, startReq, etag)
+	if err == nil {
+		err = op.WaitContext(ctx)
+	}
+
+	if err != nil {
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to start instance %q", instanceName), err.Error())
+	}
+
+	instanceStartedCheck := func() (any, string, error) {
 		st, _, err := server.GetInstanceState(instanceName)
 		if err != nil {
 			return st, "Error", err
 		}
 
+		// Beside checking for status "Running", we also need to ensure
+		// that number of procceses is higher then "0". For VMs this is
+		// only true if LXD agent is also connected to the LXD and has
+		// reported the proccess count.
+		if st.Status == api.Running.String() && st.Processes <= 0 {
+			return st, "Running (initializing)", nil
+		}
+
 		return st, st.Status, nil
 	}
+
+	// Even though op.Wait has completed, wait until we can see
+	// the instance is fully started via a new API call.
+	stateConf := &retry.StateChangeConf{
+		Target:     []string{api.Running.String()},
+		Timeout:    3 * time.Minute,
+		MinTimeout: 3 * time.Second,
+		Delay:      refInterval,
+		Refresh:    instanceStartedCheck,
+	}
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to wait for instance %q to start", instanceName), err.Error())
+	}
+
+	return nil
 }
 
-// instanceNetworkStateRefreshFunc returns function that checks
-// whether instance has received an IP address.
-func instanceNetworkStateRefreshFunc(server lxd.InstanceServer, instanceName string) retry.StateRefreshFunc {
-	return func() (any, string, error) {
+// stopInstance stops an instance with the given name. It waits wither for
+// its status to become Stopped or for the instance to be not found (in case
+// of ephemeral instances).
```

Comment:
Yep, setting `running` attribute to false is not possible for ephemeral instances. 

However, when deleting instances, we first stop them and only then delete. If instance is ephemeral, it will be deleted on stop, which we need to be aware of during deletion to not throw an error.